lib/integrations/awsoidc/create_ec2ice.go

```diff
@@ -0,0 +1,134 @@
+	if req.Cluster == "" {
+		return trace.BadParameter("cluster is required")
+	}
+
+	if req.IntegrationName == "" {
+		return trace.BadParameter("integration is required")
+	}
+
+	if req.SubnetID == "" {
+		return trace.BadParameter("subnet id is required")
+	}
+
+	if len(req.ResourceCreationTags) == 0 {
+		req.ResourceCreationTags = DefaultResourceCreationTags(req.Cluster, req.IntegrationName)
+	}
+
+	return nil
+}
+
+// CreateEC2ICEResponse contains the newly created EC2 Instance Connect Endpoint name.
+type CreateEC2ICEResponse struct {
+	// Name is the Endpoint ID.
+	Name string
+}
+
+// CreateEC2ICE describes the required methods to List EC2 Instances using a 3rd Party API.
+type CreateEC2ICEClient interface {
+	// CreateInstanceConnectEndpoint creates an EC2 Instance Connect Endpoint. An EC2 Instance Connect Endpoint
+	// allows you to connect to an instance, without requiring the instance to have a
+	// public IPv4 address. For more information, see Connect to your instances
+	// without requiring a public IPv4 address using EC2 Instance Connect Endpoint (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Connect-using-EC2-Instance-Connect-Endpoint.html)
+	// in the Amazon EC2 User Guide.
+	CreateInstanceConnectEndpoint(ctx context.Context, params *ec2.CreateInstanceConnectEndpointInput, optFns ...func(*ec2.Options)) (*ec2.CreateInstanceConnectEndpointOutput, error)
+}
+
+type defaultCreateEC2ICEClient struct {
+	*ec2.Client
+}
+
+// NewCreateEC2ICEClient creates a new CreateEC2ICEClient using a AWSClientRequest.
+func NewCreateEC2ICEClient(ctx context.Context, req *AWSClientRequest) (CreateEC2ICEClient, error) {
+	ec2Client, err := newEC2Client(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &defaultCreateEC2ICEClient{
+		Client: ec2Client,
+	}, nil
+}
+
+// CreateEC2ICE calls the following AWS API:
+// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateInstanceConnectEndpoint.html
+// It creates an EC2 Instance Connect Endpoint using the provided Subnet and Security Group IDs.
+func CreateEC2ICE(ctx context.Context, clt CreateEC2ICEClient, req CreateEC2ICERequest) (*CreateEC2ICEResponse, error) {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	createEC2ICEInput := &ec2.CreateInstanceConnectEndpointInput{
+		SubnetId:         &req.SubnetID,
+		SecurityGroupIds: req.SecurityGroupIDs,
+		TagSpecifications: []ec2types.TagSpecification{{
+			ResourceType: ec2types.ResourceTypeInstanceConnectEndpoint,
+			Tags:         req.ResourceCreationTags.ToEC2Tags(),
+		}},
+	}
+
+	ec2ICEndpoint, err := clt.CreateInstanceConnectEndpoint(ctx, createEC2ICEInput)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if ec2ICEndpoint == nil || ec2ICEndpoint.InstanceConnectEndpoint == nil || ec2ICEndpoint.InstanceConnectEndpoint.InstanceConnectEndpointId == nil {
+		return nil, trace.BadParameter("endpoint was created but failed to get its name")
```

Comment:
This error message is a bit confusing. It sounds like the endpoint failed to get its own name.

Reply:
I'll return a sentinel value like `unknown` instead of an error.

This is a side effect of the AWS SDK pattern of using pointers for everything :man_shrugging: